math/arctangent.go

```diff
@@ -1,354 +1,268 @@
+	r := stdMath.Max(0, stdMath.Ceil(0.31554321636851*stdMath.Pow(float64(ctx.Precision), 0.654095561044508)))
+	var p float64
+
+	// maxPrec is the largest precision value we can use bit shifts instead of
+	// math.Pow which is more expensive.
+	const maxPrec = 3286
+	if ctx.Precision <= maxPrec {
+		p = 1 / float64(uint64(1)<<uint64(r))
+	} else {
+		p = 1 / stdMath.Pow(2, r)
+	}
+	halved := int(stdMath.Ceil(stdMath.Log(xf/p) / stdMath.Ln2))
 
-	//if the value is already less than 1/(2^r) then halfed
+	// if the value is already less than 1/(2^r) then halfed
 	// will be negative and we don't need to apply
 	// the double angle formula because it would hurt performance
 	// so we'll set halfed to zero
-	if halfed < 0 {
-		halfed = 0
+	if halved < 0 {
+		halved = 0
 	}
 
-	if halfed > 0 {
-		x = decimal.WithPrecision(calculatingPrecision).Copy(x)
+	sq := decimal.WithContext(ctx)
+	for i := 0; i < halved; i++ {
+		ctx.FMA(sq, x0, x0, one)
+		Sqrt(sq, sq)
+		ctx.Add(sq, sq, one)
+		ctx.Quo(x0, x0, sq)
 	}
 
-	for i := 0; i < halfed; i++ {
-		xsq := decimal.WithPrecision(calculatingPrecision).Mul(x, x)
-		x = x.Quo(x, xsq.Add(Sqrt(xsq, xsq.Add(xsq, one)), one))
-	}
-
-	xSquared := decimal.WithPrecision(calculatingPrecision).Mul(x, x)
-	xSqPlus1 := decimal.WithPrecision(calculatingPrecision).Add(xSquared, one)
-	return x, xSquared, xSqPlus1, segment, halfed
+	var x2 decimal.Big
+	ctx.Mul(&x2, x0, x0)
+	var x2p1 decimal.Big
+	ctx.Add(&x2p1, &x2, one)
+	return x0, &x2, &x2p1, segment, halved
 }
 
-func getAtanA(x *decimal.Big) func(n uint64) *decimal.Big {
-	return func(n uint64) *decimal.Big {
-		return x
-	}
-}
-
-func getAtanP(precision int, xSq *decimal.Big) func(n uint64) *decimal.Big {
+func getAtanP(ctx decimal.Context, x2 *decimal.Big) SplitFunc {
+	var p decimal.Big
 	return func(n uint64) *decimal.Big {
+		// P(n) = 2n for all n > 0
 		if n == 0 {
 			return one
 		}
-
-		c2N := decimal.WithPrecision(precision).SetUint64(n)
-		c2N.Mul(c2N, two)
-		c2NXSq := c2N.Mul(c2N, xSq)
-		return c2NXSq
+		if n < stdMath.MaxUint64/2 {
+			p.SetUint64(n * 2)
+		} else {
+			ctx.Mul(&p, p.SetUint64(n), two)
+		}
+		return ctx.Mul(&p, &p, x2)
 	}
 }
 
-func getAtanB() func(n uint64) *decimal.Big {
+func getAtanQ(ctx decimal.Context, x2p1 *decimal.Big) SplitFunc {
+	var q decimal.Big
 	return func(n uint64) *decimal.Big {
-		return one
-	}
-}
+		// B(n) = (2n+1) for all n >= 0
 
-func getAtanQ(precision int, xSqPlus1 *decimal.Big) func(n uint64) *decimal.Big {
-	return func(n uint64) *decimal.Big {
-		//b(n) = (2n+1) for n >= 0
-		//most of the time n will be a small number so
-		// use the fastest method to calculate (2n+1)
-		var c2NPlus1 *decimal.Big
+		// atanMax is the largest number we can use to compute (2n + 1) without
+		// overflow.
+		const atanMax = (stdMath.MaxUint64 - 1) / 2
 		if n < atanMax {
-			c2NPlus1 = new(decimal.Big).SetUint64((n << 1) + 1)
+			q.SetUint64((2 * n) + 1)
```

Comment:
change to 2*n to n<<1 much faster

Reply:
Same as my other comment, multiplication by a power of 2 will very likely be replaced by a bit shift anyway by the compiler.

Reply:
Yep you are right.